packages/test/test-version-utils/src/versionUtils.ts

```diff
@@ -289,51 +289,89 @@ export function checkInstalled(requested: string) {
-		// if this assert fails.
-		// The v18 doc doesn't mention that export paths must start with ".", but the modern docs do:
-		// https://nodejs.org/api/packages.html#exports
-		for (const key of Object.keys(pkgJson.exports)) {
-			if (!key.startsWith(".")) {
+		if (typeof pkgJson.exports === "string") {
+			mainEntryPoint = pkgJson.exports;
+		} else {
+			// See https://nodejs.org/docs/latest-v18.x/api/packages.html#conditional-exports for information on the spec
+			// if this assert fails.
+			// The v18 doc doesn't mention that export paths must start with ".", but the modern docs do:
+			// https://nodejs.org/api/packages.html#exports
+			for (const key of Object.keys(pkgJson.exports)) {
+				if (!key.startsWith(".")) {
+					throw new Error(
+						"Conditional exports directly under 'exports' are not supported. Legacy module loading logic needs to be updated.",
+					);
+				}
+			}
+			const exportsDot = pkgJson.exports["."];
+			if (exportsDot === undefined) {
 				throw new Error(
-					"Conditional exports not supported by test-version-utils. Legacy module loading logic needs to be updated.",
+					`Package at '${pkgPath}' defined subpath exports but no '.' entry.`,
 				);
 			}
-		}
-		if (typeof pkgJson.exports === "string") {
-			primaryExport = pkgJson.exports;
-		} else {
-			const exp = pkgJson.exports["."];
-			primaryExport =
-				typeof exp === "string"
-					? exp
-					: exp.require !== undefined
-					? exp.require.default
-					: exp.default;
-			if (primaryExport === undefined) {
-				throw new Error(`Package ${pkg} defined subpath exports but no '.' entry.`);
+			const keyToCheck = moduleType === "commonjs" ? "require" : "import";
+			mainEntryPoint =
+				typeof exportsDot === "string"
+					? exportsDot
+					: exportsDot[keyToCheck].default ?? exportsDot.default;
+			if (mainEntryPoint === undefined) {
+				throw new Error(
+					`Package at '${pkgPath}' defined subpath exports but no '${keyToCheck}.default' nor 'default' entry was found under '.'.`,
+				);
 			}
 		}
 	} else {
 		if (pkgJson.main === undefined) {
-			throw new Error(`No main or exports in package.json for ${pkg}`);
+			throw new Error(
+				`No 'main' or 'exports' fields in package.json for package at '${pkgPath}'`,
+			);
 		}
-		primaryExport = pkgJson.main;
+		mainEntryPoint = pkgJson.main;
 	}
-	return import(pathToFileURL(path.join(pkgPath, primaryExport)).href);
+	return mainEntryPoint;
+}
+
+/**
+ * @internal
+ */
+export const loadPackage = async (modulePath: string, pkg: string): Promise<any> => {
+	// Because we put legacy versions in a specific subfolder of node_modules (.legacy/<version>), we need to reimplement
+	// some of Node's module loading logic here.
+	// It would be ideal to remove the need for this duplication (e.g. by using node:module APIs instead) if possible.
+	const pkgPath = path.join(modulePath, "node_modules", pkg);
+	const primaryExport = getMainEntryPointForPackage(pkgPath, "commonjs");
```

Comment:
I'm not entirely sure what to do here. Passing `"commonjs"` maintains current behavior. Might be ok-ish because older versions of (some?) packages have not been updated to use node16 module resolution and to provide ESM exports, so when those are the ones that are being loaded dynamically for compat-testing this feels correct. But not entirely sure what should happen when the "older" packages loaded dynamically start being the versions that already have node16 resolution and esm exports.

Maybe the function needs to look for an entrypoint of the specified moduleType first, but fall back to the other one before falling back to just looking for `exports['.'].default`?

Reply:
Depending on the answer here, maybe we need to be extra clear that this function is not a general solution to find the entrypoint to an arbitrary package, but specifically designed to work with FF packages which we control, know how they are evolving, and what entrypoints they do or don't provide (including potential co-existence of packages that define esm entrypoints and packages that don't within a particular release (or commit version), both of which might need to be loaded dynamically here?).

Pinging some people who might be good to keep in the loop or have touched this logic: @tylerbutler @dannimad @scottn12 @Abe27342